lints/lint_subject_contains_reserved_arpa_ip.go

```diff
@@ -0,0 +1,232 @@
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/util"
+)
+
+const (
+	// arpaTLD holds a string constant for the .arpa TLD
+	arpaTLD = ".arpa"
+
+	// rdnsIPv4Suffix is the expected suffix for IPv4 reverse DNS names as
+	// specified in https://tools.ietf.org/html/rfc1035#section-3.5
+	rdnsIPv4Suffix = ".in-addr" + arpaTLD
+	// rndsIPv4Labels is the expected number of labels for an IPv4 reverse DNS
+	// name (not counting the rdnsIPv4Suffix labels). IPv4 addresses are four
+	// bytes. RFC 1035 uses one byte per label meaning there are 4 expected labels
+	// under the rdnsIPv4Suffix.
+	rdnsIPv4Labels = 4
+
+	// rdnsIPv6Suffix is the expected suffix for IPv6 reverse DNS names as
+	// specified in https://tools.ietf.org/html/rfc3596#section-2.5
+	rdnsIPv6Suffix = ".ip6" + arpaTLD
+	// rndsIPv6Labels is the expected number of labels for an IPv6 reverse DNS
+	// name (not counting the rdnsIPv6Suffix labels). IPv6 addresses are 16 bytes.
+	// RFC 3596 Sec 2.5 uses one *nibble* per label meaning there are 16*2
+	// expected labels under the rdnsIPv6Suffix.
+	rdnsIPv6Labels = 32
+)
+
+// arpaReservedIP is a linter that errors for any well formed rDNS names in the
+// .in-addr.arpa or .ip6.arpa zones that specify an address in an IANA reserved
+// network.
+// See also: lint_subject_contains_malformed_arpa_ip.go for a lint that warns
+// about malformed rDNS names in these zones.
+type arpaReservedIP struct{}
+
+// Initialize for an arpaReservedIP linter is a NOP to statisfy linting
+// interfaces.
+func (l *arpaReservedIP) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true if the certificate contains any names that end in
+// one of the two designated zones for reverse DNS: in-addr.arpa or ipv6.arpa.
+func (l *arpaReservedIP) CheckApplies(c *x509.Certificate) bool {
+	names := append([]string{c.Subject.CommonName}, c.DNSNames...)
+	for _, name := range names {
+		name = strings.ToLower(name)
+		if strings.HasSuffix(name, rdnsIPv4Suffix) ||
+			strings.HasSuffix(name, rdnsIPv6Suffix) {
+			return true
+		}
+	}
+	return false
+}
+
+// Execute will check the given certificate to ensure that all of the DNS
+// subject alternate names that specify a well formed reverse DNS name under the
+// respective IPv4 or IPv6 arpa zones do not specify an IP in an IANA
+// reserved IP space. An Error LintResult is returned if the name specifies an
+// IP address of the wrong class, or specifies an IP address in an IANA reserved
+// network.
+func (l *arpaReservedIP) Execute(c *x509.Certificate) *LintResult {
+	for _, name := range c.DNSNames {
+		name = strings.ToLower(name)
+		var err error
+		if strings.HasSuffix(name, rdnsIPv4Suffix) {
+			// If the name has the in-addr.arpa suffix then it should be an IPv4 reverse
+			// DNS name.
+			err = lintReversedIPAddress(name, false)
+		} else if strings.HasSuffix(name, rdnsIPv6Suffix) {
+			// If the name has the ipv6.arpa suffix then it should be an IPv6 reverse
```

Comment:
Comment should reference ip6.arpa. There's no **v** in the zone name. See [link](http://www.gestioip.net/docu/ipv6_reverse_dns_delegation.html) .